src/providers/extensionBasedProvider.ts

```diff
@@ -0,0 +1,52 @@
+import { Extension } from "../extension";
+import { Provider } from "./provider";
+import { extensions } from "vscode";
+
+export class ExtensionBasedProvider extends Provider {
+    public readonly extensions: string[] = [];
+    private extensionState: Record<string, boolean> = {};
+
+    constructor(
+        extension: Extension,
+        public id = "base",
+        public priority = 0
+    ) {
+        super(extension);
+
+        this.extensionState = extensions.all
+            .filter((x) => this.extensions.includes(x.id))
+            .reduce(
+                (x, y) => {
+                    x[y.id] = y.isActive;
+                    return x;
+                },
+                {} as Record<string, boolean>
+            );
+
+        if (Object.values(this.extensionState).some((x) => !x)) this.onExtensionDeactivate();
+        else this.onExtensionActivate();
+
+        extension.context?.subscriptions.push(
+            extensions.onDidChange(() => {
+                this.updateState();
+            })
```

Comment:
`updateState` waits for the Git extension to load, which is incorrect because Git should be loaded only once. This is a Git bug for code placed after `updateState`.
```suggestion
            extensions.onDidChange(async () => {
                await this.updateState();
            })
```